app/controllers/products_controller.rb

```diff
@@ -1,26 +1,60 @@
 class ProductsController < ApplicationController
+  before_action :find_individual_product, only: [:show, :edit, :update, :retire]
+
   def index
+    @products = Product.all
   end
 
   def show
-    @product = Product.find_by(id: params[:id])
     if @product.nil?
-      head :not_found
-    end
+      flash[:error] = "Could not find product with id: #{params[:id]}"
+      redirect_to root_path, status: 302
   end
 
   def new
+    @product = Product.new
   end
 
   def create
+    product = Product.new(product_params)
+      if task.save
+        redirect_to product_path(product.id)
+      else
+        flash[:error] = "Save was unsuccessful. Try again!"
+        redirect_to root_path
+      end
   end
 
   def edit
   end
 
   def update
+    if @product.nil?
+      flash[:error] = "Could not find product with id: #{params[:id]}"
+    else
+      @product.update_attributes(product_params)
+      flash[:success] = "Product Updated"
+      redirect_to product_path(product.id)
+    end
   end
 
   def retire
+    if @product.nil?
+      flash[:error] = "Could not find product with id: #{params[:id]}"
+    elsif @product.retired == false
+      @product.retired = true
+    elsif @product.retired == true
+      @product.retired = false
+    end  
+  end
+
+  private
+
+  def find_individual_product
+    @product = Product.find_by(id: params[:id])
+  end
+
+  def product_params
+     params.require(:product).permit(:name, :price)
```

Comment:
this is what I had for product_params but let me know if this is wrong:

def product_params
    return params.require(:product).permit(
             :photo_url,
             :description,
             :name,
             :price,
             :quantity,
             :user_id,
             category_ids: [],
           )
  end

Reply:
the only thing that i don't think is a product attribute is category_ids, because it's an attribute of the join table categories_products - we should take a look at the join table before getting started today because a category_product.rb file wasn't created during the migration, and that's the only relationship that was unsuccessful when seeding the database.

Reply:
added! minus category_ids